File: T-Bank/Algorithms-and-Data-Structures-2025/5sprint-Strings/contest/A-Сравнения_подстрок/solution.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;


long long mod_pow(long long base, long long exp, long long mod) {
    long long result = 1;
    while (exp > 0) {
        if (exp % 2 == 1) {
            result = (result * base) % mod;
        }
        base = (base * base) % mod;
        exp /= 2;
    }
    return result;
}


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    const long long q = 100000009;
    const long long m = 100000003;
    string pattern;
    cin >> pattern;
    int n = pattern.size();
    vector<long long> prefixSum = {0};
    long long h = 0;
    for (char ch : pattern) {
        h = (h * q + static_cast<int>(ch)) % m;
        prefixSum.push_back(h);
    }
    int M;
    scanf("%d", &M);
    for (int i = 0; i < M; ++i) {
        int a, b, c, d;
        scanf("%d %d %d %d", &a, &b, &c, &d);
        if ((b - a + 1) != (d - c + 1)) {
            printf("No\n");
            continue;
        }
        long long q_pow = mod_pow(q, b - a + 1, m);
        long long x = (prefixSum[b] - prefixSum[a - 1] * q_pow % m + m) % m;
        q_pow = mod_pow(q, d - c + 1, m);
        long long y = (prefixSum[d] - prefixSum[c - 1] * q_pow % m + m) % m;
        cout << (x == y? "Yes" : "No") << "\n";
        
    }
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/5sprint-Strings/contest/A-Сравнения_подстрок/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    Q: int = 100000009
    M: int = 100000003
    pattern: str = sys.stdin.readline().rstrip()
    n: int = len(pattern)
    prefix_sum: list[int] = [0]
    h: int = 0
    for ch in pattern:
        h = (h * Q + ord(ch)) % M
        prefix_sum.append(h)
    m: int = int(sys.stdin.readline().rstrip())
    for _ in range(m):
        a, b, c, d = map(int, sys.stdin.readline().rstrip().split())
        if (b - a + 1) != (d - c + 1):
            sys.stdout.write('No\n')
            continue
        q_pow: int = pow(Q, b - a + 1, M)
        x: int = (prefix_sum[b] - prefix_sum[a - 1] * q_pow % M + M) % M
        q_pow: int = pow(Q, d - c + 1, M)
        y: int = (prefix_sum[d] - prefix_sum[c - 1] * q_pow % M + M) % M
        sys.stdout.write('{}\n'.format('Yes' if x == y else 'No'))


if __name__ == '__main__':
    solution()